### PR TITLE
add account_type fk in users table

### LINKED ISSUE #13751: Convert account_type in user table to a FK

### LINKED ISSUE #13752: Create migration with default account types and rate limits for on-prem/open source

### PR DESCRIPTION
**Pull request by alrocar**
Closes #13751 
Closes #13752

### CODE REVIEW

File: db/migrate/20180329095112_add_account_type_ref_to_users.rb [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+                  '{1, 1, 1}',
+                  '{1, 1, 1}')
+               RETURNING id)
+             INSERT INTO account_types (account_type, rate_limit_id)
+               VALUES (accounts[i], (select id from rate_limit));
+           END LOOP;
+         END $$;"
+    alter_table :users do
+      add_foreign_key [:account_type], :account_types, on_delete: :restrict, null: false
+      add_index :account_type
```
Comment (javitonino) [inline db/migrate/20180329095112_add_account_type_ref_to_users.rb]: Do the index first, so the foreign key creation benefits from it ;)

File: db/migrate/20180329095112_add_account_type_ref_to_users.rb [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+                                      WHERE account_type NOT IN
+                                        (SELECT account_type FROM account_types));
+         default_accounts text[] := '{\"FREE\",
+                                      \"enterprise\",
+                                      \"[DEDICATED]\",
+                                      \"ORGANIZATION USER\"}';
+
+         accounts text[] := (SELECT ARRAY_AGG(a ORDER BY a)
+                              FROM (
+                                  SELECT DISTINCT UNNEST(missing_accounts || default_accounts) AS a
```
Comment (javitonino) [inline db/migrate/20180329095112_add_account_type_ref_to_users.rb]: Have you tested this in postgres 9.5 and 10? There are some differences in behaviour in `unnest` between these two versions.

File: db/migrate/20180329095112_add_account_type_ref_to_users.rb [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+                  '{2, 2, 1}',
+                  '{1, 1, 1}',
+                  '{2, 2, 1}',
+                  '{1, 1, 1}',
+                  '{1, 1, 1}')
+               RETURNING id)
+             INSERT INTO account_types (account_type, rate_limit_id)
+               VALUES (accounts[i], (select id from rate_limit));
+           END LOOP;
+         END $$;"
```
Comment (javitonino) [inline db/migrate/20180329095112_add_account_type_ref_to_users.rb]: SQL wizardry :top:

File: db/migrate/20180329095112_add_account_type_ref_to_users.rb (hunk 2 above)
Comment (alrocar) [reply db/migrate/20180329095112_add_account_type_ref_to_users.rb]: upps, nope. Good point 👍

Comment (alrocar) [reply db/migrate/20180329095112_add_account_type_ref_to_users.rb]: Tested on both pg 9.5.2 and 10.1

**Pull request merged**